### PR TITLE
crimson/os/pg_map: allow multiple shards to create new pg mappings at the same time

### PR DESCRIPTION
**Pull request by cyx1231st**
Also:
* Better detections in case of inconsistent racings, such as:
  * The new mapping is creating towards different cores.
  * Mapping creation is racing with its eracing.
  * Multiple shards are eracing the same mapping at the same time.
* Add more logs to debug in case of unexpected issues.

Fixes: https://tracker.ceph.com/issues/64934
Fixes: https://tracker.ceph.com/issues/64009

Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
Need to validate it in sepia.

**Comment by cyx1231st** [comment]
Doesn't see the issue in the tests.

https://pulpito.ceph.com/yingxin-2024-03-21_05:50:24-crimson-rados-ci-yingxin-fix-racing-get-or-create-pg-mapping-2-distro-default-smithi/
* 7614699, 7614702: https://tracker.ceph.com/issues/62162
* The others are scrub failures

File: src/crimson/osd/pg_map.cc [hunk 1]
```diff
@@ -14,75 +14,140 @@ namespace crimson::osd {
...
-      ceph_assert_always(primary_mapping.core_to_num_pgs.size() > 0);
-      std::map<core_id_t, unsigned>::iterator core_iter;
-      if (core == NULL_CORE) {
-        core_iter = std::min_element(
-          primary_mapping.core_to_num_pgs.begin(),
-          primary_mapping.core_to_num_pgs.end(),
+    DEBUG("calling primary to add mapping for pg {} to the expected core {}",
+          pgid, core_expected);
+    return container().invoke_on(
+        0, [pgid, _core_expected=core_expected, FNAME](auto &primary_mapping) {
```
Comment (athanatos) [inline src/crimson/osd/pg_map.cc:34]: [pgid, core_expected, ...] should just work?

**Review by athanatos — APPROVED** [review]

**Comment by athanatos** [comment]
jenkins test windows

**Comment by liu-chunmei** [comment]
@cyx1231st it is also create pg on shard 0 and broadcast to others, why you name the commit : allow multiple shards to create new pg mappings at the same time

File: src/crimson/osd/pg_map.cc [hunk 2]
```diff
@@ -19,75 +19,134 @@ seastar::future<core_id_t> PGShardMapping::get_or_create_pg_mapping(
...
+          );
+          core_expected = count_iter->first;
+        } else { // core_expected != NULL_CORE
+          count_iter = primary_mapping.core_to_num_pgs.find(core_expected);
+        }
+        ceph_assert_always(primary_mapping.core_to_num_pgs.end() != count_iter);
+        ++(count_iter->second);
+        auto [insert_iter, inserted] =
+          primary_mapping.pg_to_core.emplace(pgid, core_expected);
+        assert(inserted);
```
Comment (Matan-B) [inline src/crimson/osd/pg_map.cc:74]: The fix here is to emplace the new mapping **only** if it wasn't already mapped, right? (under `find_iter == primary_mapping.pg_to_core.end()`)

File: src/crimson/osd/pg_map.cc [hunk 3]
```diff
@@ -19,75 +19,134 @@ seastar::future<core_id_t> PGShardMapping::get_or_create_pg_mapping(
...
-      if (core_expected == NULL_CORE) {
-        core_iter = std::min_element(
-          primary_mapping.core_to_num_pgs.begin(),
-          primary_mapping.core_to_num_pgs.end(),
-          [](const auto &left, const auto &right) {
-            return left.second < right.second;
+        0, [pgid, _core_expected=core_expected, FNAME](auto &primary_mapping) {
+      auto core_expected = _core_expected;
+      auto find_iter = primary_mapping.pg_to_core.find(pgid);
+      if (find_iter != primary_mapping.pg_to_core.end()) {
```
Comment (Matan-B) [inline src/crimson/osd/pg_map.cc:37]:
```
// this pgid was already mapped within primary_mapping, assert that the
// mapping is consistent and avoid emplacing once again.
if (find_iter != primary_mapping.pg_to_core.end()) {
```

**Review by Matan-B — COMMENTED** [review]
nit, commit message typo: s/eracing/erasing/

Can you please separate the actual fix from the additional asserts/logs added? If not, perhaps a comment which emphasizes where the actual concurrent new mapping occur.

File: src/crimson/osd/pg_map.cc (hunk 2 above)
Comment (Matan-B) [reply src/crimson/osd/pg_map.cc:74]: > @cyx1231st it is also create pg on shard 0 and broadcast to others, why you name the commit : allow multiple shards to create new pg mappings at the same time

@liu-chunmei

File: src/crimson/osd/pg_map.cc (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/osd/pg_map.cc:34]: I wanted to update `core_expected` but this lambda doesn't worth to become mutable.

Refer to the following logic:
```c++
auto core_expected = _core_expected;
...
  core_expected = core_found;
...
assert(core_expected != NULL_CORE);
```

Probably I can rename to make this cleaner.

File: src/crimson/osd/pg_map.cc (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/osd/pg_map.cc:74]: Right, because `primary_mapping.core_to_num_pgs` also needs to keep correct.

**Comment by cyx1231st** [comment]
> @cyx1231st it is also create pg on shard 0 and broadcast to others, why you name the commit : allow multiple shards to create new pg mappings at the same time.

This PR is allowing multiple shards to create new and consistent pg mappings at the same time, and the design needs these creation to be aligned in shard 0.

File: src/crimson/osd/pg_map.cc (hunk 3 above)
Comment (cyx1231st) [reply src/crimson/osd/pg_map.cc:37]: added

File: src/crimson/osd/pg_map.cc (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/osd/pg_map.cc]: Preserved `core_expected` and renamed to use `core_to_update` to store the updates.

**Comment by cyx1231st** [comment]
Changeset: minor renames and added some comments.

I'll proceed to merge if no further concerns.

**Comment by cyx1231st** [comment]
> Changeset: minor renames and added some comments.

I suppose there is no impact to the logic already implemented, merging based on the previous test results.

**Comment by Matan-B** [comment]
> Doesn't see the issue in the tests.
> 
> https://pulpito.ceph.com/yingxin-2024-03-21_05:50:24-crimson-rados-ci-yingxin-fix-racing-get-or-create-pg-mapping-2-distro-default-smithi/
> 
> * 7614699, 7614702: https://tracker.ceph.com/issues/62162
> * The others are scrub failures

Can you please attach the failed runs to the tracker next time? This may help with investigating :)

**Pull request merged**